### PR TITLE
chore: git mod - git context & context aware comps

### LINKED ISSUE #36811: [Task] Git mod - Migrate and update git sagas

### LINKED ISSUE #36812: [Task] Git mod - Expose APIs for sagas, action creators and store

### LINKED ISSUE #36814: [Task] Git mod - Create Git Context Provider

### LINKED ISSUE #37805: [Task] Git mod - Components - Pull button

### PR DESCRIPTION
**Pull request by brayn003**
## Description
- Adds GitContextProvider
- Adds more sagas
- Moved actions folder into store
- Minor restructuring of GitQuickActions

Fixes #36814 
Fixes #36812 
Fixes #36811 
Fixes #37805

## Automation

/ok-to-test tags="@tag.Git"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/12271500398>
> Commit: e1eb35a6c4efab61735ccc55837bdbf5bb26ccfa
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=12271500398&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Git`
> Spec:
> <hr>Wed, 11 Dec 2024 08:12:54 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
	- Introduced a new icon, `ProtectedIcon`, for enhanced visual representation.
	- Added a new component, `CtxAwareGitQuickActions`, for context-aware Git operations.
	- Implemented a new utility hook, `useStatusChangeCount`, to calculate modified entity counts.
	- Added a new component, `BranchList`, for displaying a list of branches.
	- Introduced the `GitQuickActions` component for managing various Git actions in one interface.
	- Added new actions for managing Git branch operations, including checkout, creation, and deletion.
	- Introduced new Redux sagas for handling Git operations, including checkout, commit, and fetch branches.

- **Bug Fixes**
	- Updated error handling for various Git-related actions to ensure proper state management.

- **Documentation**
	- Enhanced documentation for new components and hooks to improve developer understanding.

- **Chores**
	- Removed obsolete files and actions related to branch operations to streamline the codebase.

### CODE REVIEW

File: app/client/src/git/sagas/fetchLocalProfileSaga.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      );
+    }
+  } catch (error) {
+    yield put(
+      gitArtifactActions.fetchLocalProfileError({
+        ...basePayload,
+        error: error as string,
+      }),
+    );
+  }
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/sagas/fetchLocalProfileSaga.ts:34]: Nope error is handled outside the scope

File: app/client/src/git/sagas/deleteBranchSaga.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      );
+    }
+  } catch (error) {
+    yield put(
+      gitArtifactActions.deleteBranchError({
+        ...basePayload,
+        error: error as string,
+      }),
+    );
+  }
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/sagas/deleteBranchSaga.ts:45]: Nope error is handled outside the scope

File: app/client/src/git/requests/updateLocalProfileRequest.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+} from "./updateLocalProfileRequest.types";
+import Api from "api/Api";
+import { GIT_BASE_URL } from "./constants";
+
+export default async function updateLocalProfileRequest(
+  baseApplicationId: string,
+  params: UpdateLocalProfileRequestParams,
+): AxiosPromise<UpdateLocalProfileResponse> {
+  return Api.put(`${GIT_BASE_URL}/profile/app/${baseApplicationId}`, params);
+}
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/requests/updateLocalProfileRequest.ts:14]: Input validation is not required due to ts

File: app/client/src/git/sagas/updateLocalProfileSaga.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      yield put(gitArtifactActions.updateLocalProfileSuccess(basePayload));
+      yield put(gitArtifactActions.fetchLocalProfileInit(basePayload));
+    }
+  } catch (error) {
+    yield put(
+      gitArtifactActions.updateLocalProfileError({
+        ...basePayload,
+        error: error as string,
+      }),
+    );
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/sagas/updateLocalProfileSaga.ts:40]: This is fine

File: app/client/src/git/components/GitQuickActions/helpers/getPullButtonStatus.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+const getPullBtnStatus = ({
+  isProtectedMode = false,
+  isPullFailing = false,
+  isStatusClean = false,
+  statusBehindCount = 0,
+}: GetPullButtonStatusParams) => {
+  let message = createMessage(NO_COMMITS_TO_PULL);
+  let isDisabled = statusBehindCount === 0;
+
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/components/GitQuickActions/helpers/getPullButtonStatus.ts:24]: All props are required although, need the default values explicitly mentioned to handle component state properly

File: app/client/src/git/components/CtxAwareGitQuickActions/index.tsx [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    toggleGitConnectModal,
+    toggleGitOpsModal,
+    toggleGitSettingsModal,
+  } = useGitContext();
+
+  const isGitConnected = false;
+  const isAutocommitEnabled = true;
+  const isAutocommitPolling = false;
+  const isConnectPermitted = true;
+  const isProtectedMode = false;
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/components/CtxAwareGitQuickActions/index.tsx:24]: These will be removed later

File: app/client/src/git/store/actions/checkoutBranchActions.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+export interface CheckoutBranchInitPayload
+  extends CheckoutBranchRequestParams {}
+
+export const checkoutBranchInitAction =
+  createSingleArtifactAction<CheckoutBranchInitPayload>((state) => {
+    state.apiResponses.checkoutBranch.loading = true;
+    state.apiResponses.checkoutBranch.error = null;
+
+    return state;
+  });
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/store/actions/checkoutBranchActions.ts:14]: This is not true

File: app/client/src/git/store/actions/checkoutBranchActions.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+);
+
+export const checkoutBranchErrorAction =
+  createSingleArtifactAction<GitAsyncErrorPayload>((state, action) => {
+    const { error } = action.payload;
+
+    state.apiResponses.checkoutBranch.loading = false;
+    state.apiResponses.checkoutBranch.error = error;
+
+    return state;
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/store/actions/checkoutBranchActions.ts:31]: This is not true

File: app/client/src/git/store/actions/checkoutBranchActions.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  });
+
+export const checkoutBranchSuccessAction = createSingleArtifactAction(
+  (state) => {
+    state.apiResponses.checkoutBranch.loading = false;
+
+    return state;
+  },
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/store/actions/checkoutBranchActions.ts:21]: You are wrong

File: app/client/src/git/sagas/fetchGlobalProfileSaga.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        gitConfigActions.fetchGlobalProfileSuccess({
+          responseData: response.data,
+        }),
+      );
+    }
+  } catch (error) {
+    yield put(
+      gitConfigActions.fetchGlobalProfileError({
+        error: error as string,
+      }),
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/sagas/fetchGlobalProfileSaga.ts:28]: This is fine for now

File: app/client/src/git/sagas/updateGlobalProfileSaga.ts [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    const isValidResponse: boolean = yield validateResponse(response, true);
+
+    if (response && isValidResponse) {
+      yield put(gitConfigActions.updateGlobalProfileSuccess());
+      yield put(gitConfigActions.fetchGlobalProfileInit());
+    }
+  } catch (error) {
+    yield put(
+      gitConfigActions.updateGlobalProfileError({ error: error as string }),
+    );
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/sagas/updateGlobalProfileSaga.ts:36]: This is fine for now

File: app/client/src/git/sagas/commitSaga.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+      );
+    }
+
+    yield put(
+      gitArtifactActions.commitError({
+        ...basePayload,
+        error: error as string,
+      }),
+    );
+  }
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/sagas/commitSaga.ts:63]: This is fine for now

File: app/client/src/git/requests/connectRequest.ts [hunk 13]
```diff
@@ -4,11 +4,11 @@ import type {
...
   ConnectResponse,
 } from "./connectRequest.types";
-import type { AxiosResponse } from "axios";
+import type { AxiosPromise } from "axios";
 
 export default async function connectRequest(
   baseApplicationId: string,
   params: ConnectRequestParams,
-): Promise<AxiosResponse<ConnectResponse>> {
+): AxiosPromise<ConnectResponse> {
```
[the remark this answers is not in the document]
Comment (brayn003) [reply app/client/src/git/requests/connectRequest.ts:12]: This is already present

File: app/client/src/git/components/GitContextProvider/index.tsx [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+  const { fetchBranches } = useGitBranchesReturnValue;
+
+  useEffect(
+    function gitInitEffect() {
+      fetchBranches();
+    },
+    [fetchBranches],
+  );
+
+  // eslint-disable-next-line react-perf/jsx-no-new-object-as-prop
```
Comment (ashit-rath) [inline app/client/src/git/components/GitContextProvider/index.tsx:55]: Can we remove this?

File: app/client/src/git/components/GitQuickActions/BranchButton/index.tsx [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+      hasBackdrop
+      isOpen={isOpen}
+      minimal
+      modifiers={popoverModifiers}
+      onInteraction={onPopoverInteraction}
+      placement="top-start"
+    >
+      <Tooltip
+        content={currentBranch}
+        // eslint-disable-next-line react-compiler/react-compiler
```
Comment (ashit-rath) [inline app/client/src/git/components/GitQuickActions/BranchButton/index.tsx:89]: This as well

**Review by ashit-rath — APPROVED** [review]

**Pull request merged**